### PR TITLE
orcid: push a record to ORCID

### PR DESCRIPTION
**Pull request by szymonlopaciuk**
## Description
Push a single record to ORCID

## Checklist:

- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [x] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [x] My code follows the code style of this project.
- [x] I've added any new docs if API/utils methods were added.
- [x] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: inspirehep/modules/orcid/utils.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        string: the put-code of the updated/inserted item
+    """
+    config = load_config()
+
+    server_name = config['SERVER_NAME']
+
+    if not re.match('^https?://', server_name):
+        server_name = 'http://{}'.format(server_name)
+
+    record_api_endpoint = server_name + config['SEARCH_UI_SEARCH_API'] + recid
```
Comment (Glignos) [inline inspirehep/modules/orcid/utils.py]: You can also check if `from urlparse import urljoin` works for you.

File: inspirehep/modules/orcid/utils.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            if None will push as a new record
+
+    Returns:
+        string: the put-code of the updated/inserted item
+    """
+    config = load_config()
+
+    server_name = config['SERVER_NAME']
+
+    if not re.match('^https?://', server_name):
```
Comment (Glignos) [inline inspirehep/modules/orcid/utils.py]: Maybe this should be tested as well with a case of a server name url not containing http.

File: tests/integration/orcid/test_orcid_utils.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        ('inspirehep.net', '/api/literature/', '123', 'http://inspirehep.net/api/literature/123'),
+        ('http://inspirehep.net', '/api/literature/', '123', 'http://inspirehep.net/api/literature/123'),
+        ('https://inspirehep.net', '/api/literature/', '123', 'https://inspirehep.net/api/literature/123'),
+        ('http://inspirehep.net', 'api/literature', '123', 'http://inspirehep.net/api/literature/123'),
+        ('http://inspirehep.net/', '/api/literature', '123', 'http://inspirehep.net/api/literature/123'),
+    ]
+)
+def test_get_api_url_for_recid(server_name, api_endpoint, recid, expected):
+    result = _get_api_url_for_recid(server_name, api_endpoint, recid)
+    assert expected == result
```
Comment (iulianav) [inline tests/integration/orcid/test_orcid_utils.py]: Newline

File: .pytest_cache/v/cache/lastfailed [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+  "tests/integration/orcid/test_orcid_utils.py::test_get_api_url_for_recid[http://inspirehep.net-/api/literature-123-http://inspirehep.net/api/literature/123]": true, 
+  "tests/integration/orcid/test_orcid_utils.py::test_get_api_url_for_recid[https://inspirehep.net-/api/literature/-123-http://inspirehep.net/api/literature/123]": true
+}
```
Comment (iulianav) [inline .pytest_cache/v/cache/lastfailed]: Newline

File: inspirehep/modules/orcid/utils.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        string: API URL for the record
+    """
+    if not re.match('^https?://', server_name):
+        server_name = 'http://{}'.format(server_name)
+
+    if not api_endpoint.endswith('/'):
+        api_endpoint = api_endpoint + '/'
+
+    api_url = urljoin(server_name, api_endpoint)
+    return urljoin(api_url, recid)
```
Comment (iulianav) [inline inspirehep/modules/orcid/utils.py]: Newline

Comment (szymonlopaciuk) [reply inspirehep/modules/orcid/utils.py]: Taken care of :)

File: tests/integration/orcid/test_orcid_utils.py (hunk 3 above)
Comment (szymonlopaciuk) [reply tests/integration/orcid/test_orcid_utils.py]: This one also

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by david-caro** [comment]
ci merge please

File: inspirehep/modules/orcid/utils.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    config = load_config()
+
+    server_name = config['SERVER_NAME']
+
+    record_api_endpoint = _get_api_url_for_recid(
+        server_name, config['SEARCH_UI_SEARCH_API'], recid
+    )
+
+    hep_response = requests.get(record_api_endpoint, headers={
```
Comment (ammirate) [inline inspirehep/modules/orcid/utils.py]: I would put them in a separate function, managing corner cases too :)

File: inspirehep/modules/orcid/utils.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        'Accept': 'application/json'
+    })
+    record = hep_response.json()['metadata']
```
Comment (ammirate) [inline inspirehep/modules/orcid/utils.py]: Here I would check something like:
```python
if not record:
    Logger.error('fancy description for record {}'.format(recid))
    return
```

File: inspirehep/modules/orcid/utils.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+    bibtex_response = requests.get(record_api_endpoint, headers={
```
Comment (ammirate) [inline inspirehep/modules/orcid/utils.py]: same here eventually :P

File: inspirehep/modules/orcid/utils.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    client_secret = config['ORCID_APP_CREDENTIALS']['consumer_secret']
+
+    orcid_api = MemberAPI(client_key, client_secret)
+
+    orcid_xml = OrcidConverter(
+        record, server_name, put_code=put_code, bibtex_citation=bibtex_string
+    ).get_xml()
+
+    if put_code:
+        LOGGER.info("Pushing record #{} with put-code {}.".format(
```
Comment (ammirate) [inline inspirehep/modules/orcid/utils.py]: I would add the ORCID too, for an easy and quick check in case of errors :rocket:

**Comment by ammirate** [comment]
ci merge please

**Pull request merged**